backend/src/main/kotlin/com/example/demo/service/UserService.kt

```diff
@@ -0,0 +1,132 @@
+//    public fun findAll(): List<UserDTO> {
+//        return userRepository.findAll()
+//            .stream()
+//            .map(this::fromEntityToDTO)
+//            .collect(Collectors.toList())
+//    }
+
+    public fun create(user: User): User {
+        return userRepository.save(user)
+    }
+
+    public fun getById(id: Long): Optional<User> {
+        return userRepository.findById(id)
+    }
+
+    public fun getCardsOwnedById(id:Long, pageable: Pageable): Page<Ownership>?{
+        val exists: Optional<User> = getById(id)
+        if(exists.isPresent) {
+            val user: User = exists.get()
+            //return ownershipRepository.findByUser(user, pageable)
+            return ownershipRepository.findByUserOrderByCardCountry(user, pageable)
+        }
+        else{
+            return null
+        }
+    }
+
+    public fun updateCardsOwnedList(id: Long, card : Card): User?{
+        val exists: Optional<User> = getById(id)
+        if(exists.isPresent){
+            val user: User = exists.get()
+            //See if user already owns card
+            val ownershipValue: Ownership? = user.getCardsOwned().firstOrNull {it.getCard().getId() == card.getId()}
+            if (ownershipValue == null) {
+                val newOwnershipValue = Ownership(
+                    null,
+                    user,
+                    card,
+                    1
+                )
+                user.getCardsOwned().add(newOwnershipValue)
+                ownershipRepository.save(newOwnershipValue)
+            } else {
+                ownershipValue.setNumberOwned(ownershipValue.getNumberOwned() + 1)
+                ownershipRepository.save(ownershipValue)
+            }
+            return userRepository.save(user)
+        }
+        return null
+    }
+
+    public fun getAll(pageable: Pageable): List<User> {
+        val userEntities = userRepository.findAll(pageable)
+        val users = userEntities.map { it }
+        return users.content
+    }
+
+    fun getProgress(id: Long):String{
+        val userOpt = getById(id)
+        if (userOpt.isPresent) {
+            val cnt = userOpt.get().getCardsOwned().size
+            val percent = cnt*100.00/670
+            val decimal = BigDecimal(percent).setScale(2, RoundingMode.HALF_EVEN)
+            return "$decimal%"
+        } else {
+            throw ResponseStatusException(HttpStatus.NOT_FOUND, "User with $id not found")
+        }
+    }
+    fun getLeaders(allUser: List<User>): List<LeaderboardResponse> {
+        val listLeaders: MutableList<Pair<User, String>> = mutableListOf()
+        var counter = 1
+        for (user: User in allUser) {
+            listLeaders.add(user to getProgress(user.getId()))
+        }
+        listLeaders.sortBy { it.second }
+        listLeaders.reverse()
+        val listLeadersForPrinting: MutableList<LeaderboardResponse> = mutableListOf()
+        for (leaders: Pair<User, String> in listLeaders) {
```

Comment:
you can do this a bit more functional style
` val listLeadersForPrinting = allUser
     .map { user -> user to getProgress(user.getId()) }
     .sortedByDescending { it.second }
     .mapIndexed { index, pair -> 
        LeaderboardResponse(
            rank = index + 1,
            username = pair.first.getUsername(),
            cardsOwned = pair.first.getCardsOwned().size
        )
    }
     .toMutableList() 
`

Reply:
I cant format the code suggestion but you get the point